viz/cmd/check.go

```diff
@@ -0,0 +1,215 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/linkerd/linkerd2/pkg/healthcheck"
+	"github.com/linkerd/linkerd2/pkg/k8s"
+	"github.com/spf13/cobra"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+
+	// vizExtensionName is the name of Linkerd Viz extension
+	vizExtensionName = "linkerd-viz"
+
+	// linkerdVizExtensionCheck adds checks related to the Linkerd Viz etension
+	linkerdVizExtensionCheck healthcheck.CategoryID = vizExtensionName
+)
+
+var (
+	vizNamespace string
+)
+
+type checkOptions struct {
+	wait   time.Duration
+	output string
+}
+
+func vizCategory(hc *healthcheck.HealthChecker) (*healthcheck.Category, error) {
+
+	kubeAPI, err := k8s.NewAPI(hc.KubeConfig, hc.KubeContext, hc.Impersonate, hc.ImpersonateGroup, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	checkers := []healthcheck.Checker{}
+	checkers = append(checkers,
+		*healthcheck.NewChecker("linkerd-viz Namespace exists").
+			WithHintAnchor("l5d-viz-ns-exists").
+			Fatal().
+			Warning().
+			WithCheck(func(ctx context.Context) error {
+				return hc.CheckNamespace(ctx, vizNamespace, true)
+			}))
+
+	checkers = append(checkers,
+		*healthcheck.NewChecker("linkerd-viz ClusterRoles exist").
+			WithHintAnchor("l5d-viz-sc-exists").
+			Fatal().
+			Warning().
+			WithCheck(func(ctx context.Context) error {
+				return healthcheck.CheckClusterRoles(ctx, kubeAPI, true, []string{fmt.Sprintf("linkerd-%s-prometheus", vizNamespace), fmt.Sprintf("linkerd-%s-tap", vizNamespace)}, "")
+			}))
+
+	checkers = append(checkers,
+		*healthcheck.NewChecker("linkerd-viz ClusterRoleBindings exist").
+			WithHintAnchor("l5d-viz-sc-exists").
+			Fatal().
+			Warning().
+			WithCheck(func(ctx context.Context) error {
+				return healthcheck.CheckClusterRoleBindings(ctx, kubeAPI, true, []string{fmt.Sprintf("linkerd-%s-prometheus", vizNamespace), fmt.Sprintf("linkerd-%s-tap", vizNamespace), "linkerd-web"}, "")
+			}))
+
+	//TODO: add tap webhook certs check
+
+	checkers = append(checkers,
+		*healthcheck.NewChecker("collector config map exists").
+			WithHintAnchor("l5d-viz-oc-cm-exists").
+			Warning().
+			WithCheck(func(ctx context.Context) error {
+				// Check for viz Service Account
+				_, err = kubeAPI.CoreV1().ConfigMaps(vizNamespace).Get(ctx, "collector-config", metav1.GetOptions{})
+				if err != nil {
+					return err
+				}
+				return nil
+			}))
+
+	checkers = append(checkers,
+		*healthcheck.NewChecker("viz extension pods are running").
+			WithHintAnchor("l5d-viz-collector-running").
+			Warning().
+			WithRetryDeadline(hc.RetryDeadline).
+			SurfaceErrorOnRetry().
+			WithCheck(func(ctx context.Context) error {
+				pods, err := kubeAPI.GetPodsByNamespace(ctx, vizNamespace)
```

Comment:
Does this just check that the pods which exist in the namespace are running?

Should it also check that the correct pods exist in the first place?  Or perhaps that each of the expected deployments has a non-zero number of ready replicas?

Reply:
Checking for the right pods to exist seemed to be the right thing.

 We can also add a another check for replica sets. WDYT?